Update functions/constants/events and fix other issues.

Sei-Lisa opened the pull request:
### THIS COMMIT IS INCOMPLETE. See below.

The included fixes (per the commit message) are as follows:

- 'function' does not have any meaning in LSL.
- 'return' had a space which caused problems.
- Added 'event', 'print' as reserved words, even though they aren't flow control keywords. They may need to be removed from here; see below.
- Remove special treatment of \uXXXX string escape which does not work in LSL.
- Make '0x' prefix not accept an empty string after it.
- Make '\.' be the regex for quoted characters.
- Accept any kind of spacers, including newlines, between the identifier and the parentheses, to take it as a user function.
- (Edit - I forgot that in the commit message) Exponents in scientific notation accept a plus sign.

There are, however, a bunch of outstanding issues pending.

- Multi-line comments are not handled. They can't be nested, i.e. `/*a/*b*/c*/` is taken as `c*/` and will cause a syntax error (to answer the question in #1).
- Strings are still not highlighted, no idea why.
  - LSL strings are similar to C strings, with the difference that they can be multi-line.
  - The only escapes interpreted are `\n` and `\t`. Every other escape just takes the next character literally, including `"` and `\` as expected, but also any other letter, number or symbol.
  - If used within a string, a `\` at the end of a line causes problems due to bugs in the LSL lexical analyzer. Out of strings, `\` are ignored no matter the placement.
- I haven't removed `constant.language.lsl` because I don't know if removing it causes problems, but I don't think that it's necessary to have both `constant.language.lsl` and `support.constant.lsl`. If both are necessary, the current list seems appropriate, as `TRUE` and `FALSE` receive special treatment by the grammar (though that's a little known fact and may confuse rather than help). If one is removed, `TRUE` and `FALSE` will need to be re-added to the other regex, which currently doesn't include them.
- I don't know how to add more categories, but most other syntax highlighters that accept LSL, have up to six more keyword categories:
  - Deprecated functions:
    llCloud
    llMakeExplosion
    llMakeFire
    llMakeFountain
    llMakeSmoke
    llRemoteDataSetRegion
    llSound
    llSoundPreload
    llXorBase64Strings
    llXorBase64StringsCorrect
  - Unimplemented functions:
    llClearExperiencePermissions
    llCollisionSprite
    llGetExperienceList
    llPointAt
    llRefreshPrimURL
    llReleaseCamera
    llRemoteLoadScript
    llSetPrimURL
    llStopPointAt
    llTakeCamera
  - GodMode functions (they require the user to have "god powers" and don't work for most)
    llGodLikeRezObject
    llSetInventoryPermMask
    llSetObjectPermMask
  - Deprecated constants:
    PSYS_SRC_INNERANGLE
    PSYS_SRC_OUTERANGLE
    VEHICLE_FLAG_NO_FLY_UP
    ATTACH_RPEC
    ATTACH_LPEC
    LAND_SMALL_BRUSH
    LAND_MEDIUM_BRUSH
    LAND_LARGE_BRUSH
    DATA_RATING
    PRIM_CAST_SHADOWS
    PRIM_MATERIAL_LIGHT
  - Unimplemented constants (the constants exist but they serve no purpose, even if they imply one).
    PERMISSION_REMAP_CONTROLS
    PERMISSION_RELEASE_OWNERSHIP
    PERMISSION_CHANGE_JOINTS
    PERMISSION_CHANGE_PERMISSIONS
    PSYS_SRC_OBJ_REL_MASK
  - Unused keywords. These are `event` and `print`. They are reserved keywords, not available for user identifiers, therefore they should be highlighted. `event` is tokenized by the lexical analyzer but is unused in the grammar, and `print` is a specially handled function that does nothing at all. **If a category is created for them, they should be removed from the 'flow control' regex** where I placed them.

**The above keywords are not included in this commit.** A few of the deprecated functions are still in use, so they need to be highlighted. I can add them if I can have a base with the categories prepared in advance and the regular expressions ready to be filled in. However it should be straightforward for anyone to fill them in based on the above lists.

If it's absolutely necessary to simplify, It is possible to merge e.g. Unimplemented+Deprecated into a single Deprecated category, and GodMode keywords can be omitted, as they don't compile for mere mortals.

Sei-Lisa commented:
Apart from learning the full XML format expected and what category names are acceptable and what are not, and submitting a PR with the full set of changes necessary included, is there anything else I can do to move this forward? Provide the full list of keywords separated by category perhaps? Provide the link to the [regex optimizer](http://search.cpan.org/~dankogai/Regexp-Optimizer-0.23/lib/Regexp/Optimizer.pm) used to create that regex? Anything else?

infininight commented:
I made some quick improvements on the way to reviewing this pull request. I can probably largely pull it in as-is, checking things as I go. We would prefer if each commit be a distinct change such as mine, do you want to break them out or should I just part them into my own commits?

Sei-Lisa committed:
'function' is not a type; 'quaternion' is an alias of 'rotation'

Sei-Lisa committed:
Highlight 'event' and 'print' as reserved words.

These are deprecated and not flow control words, so maybe they deserve a new category instead.

Sei-Lisa committed:
'0x' has to be followed by at least one hex char.

Sei-Lisa committed:
Exponents in scientific notation admit '+' signs.

e.g. 1.0E+2 is just fine.

Sei-Lisa committed:
Accept any kind whitespace and not just spaces, for user functions.

Previously, this:
   myfunc ();
was a user function, but this:
   myfunc
   ();
was not. This change tries to correct that.

Sei-Lisa committed:
Update list of events to latest version as of 2017-09-06.

Sei-Lisa committed:
Update list of library functions, but some are missing.

This is the list as of 2017-09-06, with the exception of the following:

- Deprecated functions (they work but they should not be used, as there are replacements):
  llCloud
  llMakeExplosion
  llMakeFire
  llMakeFountain
  llMakeSmoke
  llRemoteDataSetRegion
  llSound
  llSoundPreload
  llXorBase64Strings
  llXorBase64StringsCorrect
- Unimplemented functions (these do nothing):
  llClearExperiencePermissions
  llCollisionSprite
  llGetExperienceList
  llPointAt
  llRefreshPrimURL
  llReleaseCamera
  llRemoteLoadScript
  llSetPrimURL
  llStopPointAt
  llTakeCamera
- God-Mode functions (only Linden Lab employees with certain powers can use them, otherwise they don't compile at all):
  llGodLikeRezObject
  llSetInventoryPermMask
  llSetObjectPermMask

As of this writing, at least one of these functions, llXorBase64StringsCorrect, is still current (not deprecated) in OpenSim.

Sei-Lisa committed:
Update the list of constants, but some are still missing.

TRUE and FALSE are in the constant.language.lsl category alone. They are treated by the compiler slightly differently (they can't have a minus sign if used in the globals section, while the others can). The rest are in the support.constant.lsl category. This distinction may be artificial and unnecessary.

This is the list of constants as of 2017-09-06 with these exceptions:

- Deprecated constants (because they have a misleading name, or for other reasons)
  PSYS_SRC_INNERANGLE
  PSYS_SRC_OUTERANGLE
  VEHICLE_FLAG_NO_FLY_UP
  ATTACH_RPEC
  ATTACH_LPEC
  LAND_SMALL_BRUSH
  LAND_MEDIUM_BRUSH
  LAND_LARGE_BRUSH
  DATA_RATING
  PRIM_CAST_SHADOWS
  PRIM_MATERIAL_LIGHT
- Unimplemented constants (they exist but they serve no purpose, even if they imply one):
  PERMISSION_REMAP_CONTROLS
  PERMISSION_RELEASE_OWNERSHIP
  PERMISSION_CHANGE_JOINTS
  PERMISSION_CHANGE_PERMISSIONS
  PSYS_SRC_OBJ_REL_MASK

Sei-Lisa commented:
Thank you very much for reviewing and for your changes. I've force-pushed a rebased version with the changes separated in different commits. There is a change that arguably could be split into two, namely the first commit (remove `function` and add `quaternion`). I'll do that if requested.

I've split the long expressions into separate lines by initial. Every library function starts with `ll`, so I've put that part in the same line as the `\b(` to save an indentation.

This is still not ready to go live due to the missing keywords. I can probably figure out how to add the missing categories, as long as I'm told what category names are acceptable. It would be preferable if `event` and `print` also get a category, as these are invalid.

You were right, strings are highlighted but I couldn't see the difference in my monitor between the dark grey of the normal text and the dark blue of the strings. Taking a screenshot and zooming showed my mistake. Apologies about that.

infininight commented on Syntaxes/Second Life LSL.tmLanguage:
```diff
@@ -109,7 +109,7 @@
 		<dict>
 			<key>match</key>
-			<string>\b\w+ *(?=\()</string>
+			<string>\b\w+\s*(?=\()</string>
```
This is still a good change as it would allow tab characters here, but it doesn't do what you think. Each regex rule can only see a single line at a time, so matching a function with parens on the next line can't work. Rules can span multiple lines with begin/end but each regex match is a single line.

infininight reviewed (commented):
So the way to mark deprecated items is with `invalid.deprecated.*` this will mark the items deprecated though with suitable theming implications. Is this objectionable with regards to OpenSim?

Sei-Lisa commented:
I think OpenSim users are aware of that. Even if it shares most of the functions, it has a lot of new functions that aren't present here and it lacks a few that are present in SL. Perhaps it deserves a separate bundle, should anyone be willing to add it.

Thanks for the info on `\s`. It's still good enough, I suppose. It's not that common that anyone would separate the function name from the parenthesis with a newline, anyway.

So please hold this PR until I push two more changes, one adding more extensions (.lslp, .lslm, .lslo) and the other adding all the missing functions and constants under a common `invalid.deprecated.lsl` if that's OK. If it's possible to separate it further, e.g. `invalid.deprecated.godmode.lsl` and `invalid.deprecated.unimplemented.lsl` please let me know, as that would be ideal. Otherwise they will all be under the common `invalid.deprecated.lsl` which, though not ideal, would be acceptable.

infininight commented:
Based on your previous comment this is where I'd put them last night:

https://gist.github.com/infininight/a3facce565395ef570258f0ccccc0e17

`invalid.deprecated` is for items that are still valid just not recommended, `invalid.illegal` is for items that should never be used, in this case items that are reserved for future use. The god-mode functions are valid even if of limited use so they get a standard but extended scope.

And yes it is perfectly fine to extend the scopes like this, the first two parts are standardized, then anything descriptive/etc next, then always ending in the language.

Sei-Lisa committed:
Add some file extensions.

.lslm is for LSLForge modules.
.lslp is for LSLForge (formerly LSLPlus) scripts.
.lslo is for scripts optimized by LSL-PyOptimizer.

Sei-Lisa committed:
Add deprecated/unimplemented functions.

Sei-Lisa committed:
Add deprecated constants.

Sei-Lisa commented:
Thanks for your guidance. I've added the deprecated functions following your scheme. I've merged the deprecated and unimplemented constants into one single category, `invalid.deprecated.constant.lsl` which I hope is acceptable.

It's three commits instead of the two I announced due to splitting the keywords into functions/constants.

I've noticed a small inconsistency in the naming scheme: lsl is not always at the end. This only affects `constant.language.lsl.events`, which, by the way, sounds like an inadequate category. Events are special functions.

Besides that, I'd now call it complete and I think it fully addresses #1.

infininight commented:
OK, I've pulled all of these in along with a few other changes. Have not deployed publicly yet, please check to make sure everything looks ok there.

\* Changes not live on Github so if you can't test in TextMate you can use Lightshow and paste the entire grammar and some sample code:

https://github-lightshow.herokuapp.com

Sei-Lisa commented:
There are still a couple issues; one I forgot in #1 and the other as a result of 356910d.

Library functions (all that start with `ll`) can be valid as local variable identifiers, therefore I suggest this change at the end of each expression (this is not an actual diff, it's just to indicate what to change):
```diff
- ))\b
+ ))(?=\s*\()
```

As for numbers:
- Numbers like `1e-4` are not integers; they are floats.
- Floats that have a dot (e.g. `1.3`, `1.`, `.5`, `1.5e+3`) can have an 'f' or 'F' as a suffix. Floats that don't (e.g. 1e3) can't. Not sure if this is worth implementing, though, as that's a leftover from the C lexer they used as a basis. I ignored it when the expression was different, but I saw this separation as a chance to implement it.
- Minus signs are part of the number in very specific cases, hard to distinguish:
  - In globals, `integer a = -1;` the minus is part of the number. In regular code, with the same line, the minus is not part of the number, but a separate operator.
  - Immediately after a typecast, a minus sign is part of the number, e.g. `(float)-1`.
  - In all other cases, it's a separate operator.

However, the way it's written, the signs in `x+1` and `x-1` are treated differently (the first as an operator and the second as part of the number), which is inconsistent. I'd suggest removing the sign from the number and treating it as an operator always.

Another minor issue is that strings admit an `L` prefix, but that again is a leftover of the C lexer it's based on.

I've created this test:

http://lsl.project.zone/paste/view/a1a32cbd

(you can see it there as highlighted by GeSHi, for comparison; use View Raw to download). I ran it through Lightshow and apart from these issues I didn't see anything wrong. Perhaps the escape characters in strings are not clearly distinguished, but I guess that has to do with the theme.

Lastly, another minor issue is that events are not constants; they are special functions.

Sei-Lisa commented:
Just noticed another minor issue when inspecting the Lightshow output again: the `@` and `[...]` are not highlighted as symbols. `@` is a label prefix; `[...]` are used for lists (which are somewhat similar to Python's lists).

Sei-Lisa commented:
Sorry for the spam. While reviewing, you can check the validity of code here: [LSL-PyOptimizer Online](http://lsl.project.zone/lsl-pyoptimizer/online.php); if it's just an expression, you can use the [LSL calculator](http://lsl.project.zone/lslcalc/lslcalc.php) instead.

I've updated the testcase a bit. New link: http://lsl.project.zone/paste/view/91f506fe

infininight commented:
- Library functions (re: `))(?=\s*\()`): This change would require the parenthesis which currently we don't, this isn't desired? Whereas the parenthesis is currently allowed just fine no?
- Negative numerics: Fixed
- `1e-4` floats: What is the distinction that tips a number into being a float here? Is there a BNF of the grammar somewhere?
- `L"string"`: Is it something that should be fixed since it doesn't actually serve a purpose?
- `@`: We would highlight this not separately but with a punctuation scope as part of a label match which doesn't currently exist.
- `[]`: Same way really, would be part of a punctuation scope and a larger match which isn't really matched in a proper way currently.

Sei-Lisa commented:
- On the need of parentheses: It's mostly academic as no one would use an existing library function's name for a local variable or label, so I'm OK with not adding it.
- `1e-4` floats: If it is numbers, optionally with one dot, followed by an exponent, or numbers with a mandatory dot, it's a float. If it's only numbers, or `0x` + hex chars, it's an integer. Yes, old versions of the flex and bison input files are available (the only difference expected is the keywords). The scanner is here: https://bitbucket.org/lindenlab/viewer-release/src/01be5a9c270ff3564d9e91096d254b4d6532eb91/indra/lscript/lscript_compile/indra.l?at=default and the parser, which is probably unnecessary in this case, is here: https://bitbucket.org/lindenlab/viewer-release/src/01be5a9c270ff3564d9e91096d254b4d6532eb91/indra/lscript/lscript_compile/indra.y?at=default . The float section is summarized as follows:
    - `{N}+{E}`
    - `{N}*"."{N}+({E})?{FS}?`
    - `{N}+"."{N}*({E})?{FS}?`
  where `{N}` is `[0-9]`, `{E}` is `[Ee][+-]?{N}+` and `FS` is `(f|F)`. I'm OK with leaving out the `F` as no one uses it, but when it's used it causes wrong highlighting.
- `L"string"` - I'm OK with leaving it out.
- `@` labels: Fine with me. Label statements are always `"@" identifier` (they are not a prefix like in C, they are a statement on their own).
- `[...]`: Given that they admit general expressions inside, including strings that can contain unmatched `[` or `]` and comments, that sounds a bit overkill to me. But if you think it's easy, no problem with me.

infininight commented:
OK, pushed an update which I think covers everything.

- Improved punctuation scopes for arrays as well as `{}` and `()` sections.
- Numerics have been overhauled a bit.
- Identifiers have been strict-ified.
- Labels are now matched in both sides of the usage. You may object to the scope used here but the goal when we do syntax highlighting is to be helpful to the author not to be strictly correct in naming. Scoping as a constant variant allows them to quickly jump out in code.

Sei-Lisa commented:
Looks awesome now. On inspection I wondered about the \B in the label; I guess it's to prevent cases like x@y; to be highlighted as labels.

Tested in Lightshow and very pleased with the result. Thanks a lot. :+1:

infininight commented:
Correct the `\B` prevents the `@` from being matched mid-word, the inverse of the `\b` used before other identifiers.

I've deployed this for TextMate users. For Github it will start being used next linguist deploy, these happen about every six weeks so the time until you see it will vary depending on when in the cycle we are.

Thanks!
